File: web/src/models/dayKey.ts
/* Day keys ported from Bulk/Support/DayKey.swift. All diary grouping uses
   local-timezone calendar days, serialized as "YYYY-MM-DD". Arithmetic goes
   through local Date construction so DST transitions can't skip days. */

function pad(n: number): string {
  return n < 10 ? `0${n}` : String(n);
}

/** Local-timezone day key for a Date. */
export function dayKeyOf(date: Date): string {
  return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}`;
}

export function todayKey(now: Date = new Date()): string {
  return dayKeyOf(now);
}

/** Noon local time on the keyed day — a DST-safe anchor for date math. */
export function dayKeyToDate(key: string): Date {
  const [y, m, d] = key.split("-").map(Number);
  return new Date(y, m - 1, d, 12);
}

export function shiftDayKey(key: string, days: number): string {
  const date = dayKeyToDate(key);
  date.setDate(date.getDate() + days);
  return dayKeyOf(date);
}

/** Whole calendar days from `a` to `b` (positive when b is later). */
export function dayKeyDiff(a: string, b: string): number {
  const ms = dayKeyToDate(b).getTime() - dayKeyToDate(a).getTime();
  return Math.round(ms / 86_400_000);
}

/** "Today", "Yesterday", "Tomorrow", or a medium formatted date. */
export function dayKeyDisplayName(key: string, now: Date = new Date()): string {
  const today = todayKey(now);
  if (key === today) return "Today";
  if (key === shiftDayKey(today, -1)) return "Yesterday";
  if (key === shiftDayKey(today, 1)) return "Tomorrow";
  return dayKeyToDate(key).toLocaleDateString(undefined, {
    month: "short",
    day: "numeric",
    year: "numeric",
  });
}

File: Bulk/Support/DayKey.swift
import Foundation

/// All diary grouping uses local-timezone calendar days. Entries store a full
/// timestamp plus this normalized "start of day" key.
enum DayKey {
    static func of(_ date: Date, calendar: Calendar = .current) -> Date {
        calendar.startOfDay(for: date)
    }

    static func today(calendar: Calendar = .current) -> Date {
        calendar.startOfDay(for: Date())
    }

    static func shifted(_ dayKey: Date, by days: Int, calendar: Calendar = .current) -> Date {
        calendar.date(byAdding: .day, value: days, to: dayKey) ?? dayKey
    }

    static func isToday(_ dayKey: Date, calendar: Calendar = .current) -> Bool {
        calendar.isDateInToday(dayKey)
    }

    /// "Today", "Yesterday", "Tomorrow", or a medium formatted date.
    static func displayName(for dayKey: Date, calendar: Calendar = .current) -> String {
        if calendar.isDateInToday(dayKey) { return "Today" }
        if calendar.isDateInYesterday(dayKey) { return "Yesterday" }
        if calendar.isDateInTomorrow(dayKey) { return "Tomorrow" }
        return dayKey.formatted(date: .abbreviated, time: .omitted)
    }
}
